cumulusci/core/source/github.py

```diff
@@ -0,0 +1,106 @@
+import os
+
+from cumulusci.core.github import get_github_api_for_repo
+from cumulusci.core.github import find_latest_release
+from cumulusci.utils import download_extract_github
+
+
+class GitHubSource:
+    def __init__(self, project_config, spec):
+        self.project_config = project_config
+        self.spec = spec
+        self.url = spec["github"]
+        if self.url.endswith(".git"):
+            self.url = self.url[:-4]
+
+        repo_owner, repo_name = self.url.split("/")[-2:]
+        self.repo_owner = repo_owner
+        self.repo_name = repo_name
+
+        self.gh = get_github_api_for_repo(
+            project_config.keychain, repo_owner, repo_name
+        )
+        self.repo = self.gh.repository(self.repo_owner, self.repo_name)
+        self.resolve()
+
+    def __repr__(self):
+        return f"<GitHubSource {str(self)}>"
+
+    def __str__(self):
+        s = f"GitHub: {self.repo_owner}/{self.repo_name}"
+        if self.description:
+            s += f" @ {self.description}"
+        if self.commit != self.description:
+            s += f" ({self.commit})"
+        return s
+
+    def __hash__(self):
+        return hash((self.url, self.commit))
+
+    def resolve(self):
+        """Resolve a github source into a specific commit.
+
+        The spec must include:
+        - github: the URL of the github repository
+
+        The spec may include one of:
+        - commit: a commit hash
+        - ref: a git ref
+        - branch: a git branch
+        - tag: a git tag
+
+        If none of these are specified, CumulusCI will look for the latest release.
+        If there is no release, it will use the default branch.
+        """
+        ref = None
+        if "commit" in self.spec:
+            self.commit = self.description = self.spec["commit"]
+            return
+        elif "ref" in self.spec:
+            ref = self.spec["ref"]
+        elif "tag" in self.spec:
+            ref = "tags/" + self.spec["tag"]
+        elif "branch" in self.spec:
+            ref = "heads/" + self.spec["branch"]
+        if ref is None:
+            release = find_latest_release(self.repo, include_beta=False)
+            if release:
+                ref = "tags/" + release.tag_name
+            else:
+                ref = "heads/" + self.repo.default_branch
+        self.description = ref[6:] if ref.startswith("heads/") else ref
+        self.commit = self.repo.ref(ref).object.sha
+
+    def fetch(self, path=None):
+        """Fetch the archive of the specified commit and construct its project config."""
+        # To do: copy this from a shared cache
```

Comment:
did you want to address this TODO on this iteration of the functionality?

Reply:
I'm intentionally leaving this one for later. It's a nice-to-have for reducing network traffic a bit, but isn't critical and shouldn't change functionality.